onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -0,0 +1,428 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "common_subexpression_elimination.h"
+#include "core/graph/graph_utils.h"
+
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace onnxruntime {
+
+namespace {
```

Comment:
A big comment with an overview of what is being done and how, and example/s would be really helpful here. 
Including a link to something like [https://en.wikipedia.org/wiki/Common_subexpression_elimination](https://en.wikipedia.org/wiki/Common_subexpression_elimination) would be good as well.